### ISSUE TITLE
new/try_new with mut owner as lambda parameter

### ISSUE THREAD

**Issue by amdjachenko** [Voultapher/self_cell#52]
currently new and try new expects a lambda that accepts a ref on the owner to produce a dependent object but I need to have mut reference on the owner to produce the dependent object. I use rusqlite connection as the owner and inner self_cell as the dependent object. I use statement as the inner owner to produce dependent rows. So I need the statement to be mut in order to have rows.

```rust
        use sql::Rows;
        self_cell::self_cell! {
            struct RowsCell<'a> {
                owner: sql::Statement<'a>,
                #[covariant]
                dependent: Rows,
            }
        }

        self_cell::self_cell! {
            struct StatementCell {
                owner: sql::Connection,
                #[covariant]
                dependent: RowsCell,
            }
        }
                ----------------------
                self.stmt = Some(StatementCell::try_new(conn, |conn| {
                    RowsCell::try_new(conn.prepare(query)?, |stmt| {
                        (unsafe { &mut *(stmt as *const sql::Statement as *mut sql::Statement) })
                            .query(sql::params_from_iter(values))
                            .and_then(|mut r| {
                                r.advance()?;
                                Ok(r)
                            })
                    })
                })?);
```

as you can see I simply cast ref to mut ref and everything worked fine until I updated my toolchain to the latest version, and now I have this error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`

It seems to me the restriction to have ref owner in new/try_new is artificial and doesn't have any real reason. Am I wrong? If not would you please consider to lift the restriction?

**Comment by Voultapher** [comment]
I previously https://github.com/Voultapher/self_cell/pull/36 thought about supporting this use case, but decided it's not a good fit for self_cell. Ouroboros supports this use case.

Please let me know if this answers your question.

**Comment by Voultapher** [comment]
I've not heard back from you. So I'm assuming this issue is addressed. If you disagree, feel free to re-open the issue.

**Comment by amdjachenko** [comment]
Sorry, I lost this page in my browser and didn't see your answer before. I read the discussion in 36 issue. I see why you opposed to this change. For now to resolve my issue I've just split one line casting from ref to mut ref into two step, first acquire const pointer in one line and then casting it to mut pointer and getting the mut ref in the second line so I don't having any blocking issue. 

What about making unsafe_new/unsafe_try_new with mut owner? This way it both solves my (and other users) issue and clearly communicate the risk and what exactly should be checked (no mut ref on owner in dependent struct). In my case its safe(assuming rusqlite is safe) as I do nothing that is not already allowed by rusqlite.

**Comment by Voultapher** [comment]
Regarding your workaround, that is UB. Both stacked borrows and tree borrows agree:

```rust
fn main() {
    let mut owning_string = String::from("some string content");

    let shared_ref: &String = &owning_string;
    let shared_ptr: *const String = shared_ref;
    let unique_ptr: *mut String = shared_ptr as *mut String;

    // SAFETY: This is UB.
    unsafe {
        (*unique_ptr).pop();
    }
}
```

Please use ouroboros if you can't structure your program in another way.

Regarding the suggested fix. It is incompatible with "**Safe-to-use** proc-macro-free self-referential structs in stable Rust." crate slogan. I explicitly don't wan't to ask users to write unsafe code, as by and large they invariably get it wrong.

**Comment by amdjachenko** [comment]
sure, in your example this is UB, but my case is using rusqlite in a very specific case with your library which restricts mut owner just because somebody may keep mut ref owner in dependant by mistake (My use case don't requires this at all). 

usqlite requires mut ref statement not for mutating internal data (which is held as just ref which in turn held a pointer to C sqlite statement) but for leverage rust borrow checker to prevent making more than 1 rows abtraction, more specifically it doesn't prevent this but reset all previously created rows and as I said I do nothing with rusqlite that is not already allowed, so that is just details and irrelevant to the suggestion. The main point is there are a ligit/sound set of cases where using unsafe version of new with mut owner would solve artificial problems. Sure it needs to be documented to specify what case is ok and what is UB.

In any way, I don't see how explicit additional unsafe functions contradicts to "Safe-to-use proc-macro-free self-referential structs in stable Rust" it would explicitly marked as unsafe **additional** functionality, so all your promises will be held as now, or you want to say that "safe-to-use" part == unsafe is totally banned for any purpose? Ok, in this case, I'll just continue to use double casting as before untill I found a real reason to swicth to ouroboros.

**Comment by Voultapher** [comment]
My example is UB even without mutation. If we replace `(*unique_ptr).pop();` with `let x: &mut String = &mut *unique_ptr;` Miri fails and prints, see https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=da7c85a1e618e9a85d0116d2d821e55a:

```
error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
```

Your code invokes UB.

**Comment by Voultapher** [comment]
Also did you consider putting the owner inside a `RefCell`?

**Comment by amdjachenko** [comment]
> Your code invokes UB.

I am from c++ world, it is not so simple to scare me by naming something is UB where it is literally everywhere but we all still alive. All UBs are different and many need to be taken very seriously but 
> casting `&T` to `&mut T` is undefined behavior, even if the reference is unused

will not cause me to worry, I left responsibility about this to people who wrote this message. Don't say wolf where there is no one - a good strategy to build trustworthy relationships.

> Also did you consider putting the owner inside a RefCell?

I tried, borrow_mut() returns RefMut guard object, not the mut ref owner, so it alows me to call mutation methods on owner but only in the lambda scope. As I said Rows abstraction (pure rust thing that doesn't exist in C sqlite) keeps ref on stmt internally (to prevent creation any other rows while this instance alive) and this causing an error  
> cannot return value referencing local variable `stmt_guard` returns a value referencing data owned by the current function

```rust
                self.stmt = Some(StatementCell::try_new(conn, |conn| {
                    RowsCell::try_new(std::cell::RefCell::new(conn.prepare(query)?), |stmt| {
                        let mut stmt_guard = stmt.borrow_mut();
                        stmt_guard.query(sql::params_from_iter(values))
                            .and_then(|mut r| {
                                r.advance()?;
                                Ok(r)
                            })
                    })
                })?);
```

So basically I am back to the same place why I need to use your library in the first place. This problem caused by not the best rusqlite API design, instead of keeping reference on the stmt inside the rows, rows constructor could consume stmt (by value) to prevent creation other rows on the same statement and then return it back when no more rows available, basically using values everywhere always better then using references but I can't influence this design change.

**Comment by NobodyXu** [comment]
> > casting `&T` to `&mut T` is undefined behavior, even if the reference is unused
> 
> will not cause me to worry, I left responsibility about this to people who wrote this message. Don't say wolf where there is no one - a good strategy to build trustworthy relationships.

If you do that with optimization enabled, then mis-compilation might happen.

The compiler might assume that it can cache the value it reads.

**Comment by Voultapher** [comment]
Look I don't know what to tell you @amdjachenko. If you don't respect the language rules I can't help you.

**Comment by amdjachenko** [comment]
With all respect to your attempts to help me and point on language rules, you know you kind of fooled by the language developers. Rules are not laws of physics or math logic they are people rules == controversial by nature.

> if you do that with optimization enabled, then mis-compilation might happen.
The compiler might assume that it can cache the value it reads.

I see exactly 0 problems, if it caches the value and (remember reference is unused) no dereference of mut ref happens -> no change of that variable can happen (via that mut ref). If it caches the value and the ref is not used (but mut ref is used only) again no issue. So if you do that trick you should discard ref and use only mut ref but never ever use both of them.

I understand motiviation of language developers to call that UB but I don't share their approach because this is not true in all the cases.

**Comment by NobodyXu** [comment]
@amdjachenko In that case where you know what you are doing, I recommend you to build your own safe abstraction using raw pointers.

**Comment by NobodyXu** [comment]
You can create arbitrary pointers, mutable or immutable, as long as you use them correct (no unsynchronised concurrent write/read, no out-of-bound/invalid use, interleaved write and read, etc) then it's considered as safe.

The existence of pointers in Rust is for you to create new safe abstractions, which sounds like this is the scenario you are in, where existing safe abstraction isn't good enough for you.
